rviz_simulator/src/camera_calibration_optimizer.cpp

```diff
@@ -0,0 +1,636 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Christopher Sahadeo
+ *
+ * Ceres optimizer for a bundle adjustment camera calibration problem
+ * Considers the reprojection error to optimize the following parameters:
+ *    world_T_camera
+ *    world_T_target
+ *    camera_intrinsics
+ *
+ */
+
+#include "rviz_simulator/camera_calibration_optimizer.h"
+
+namespace camera_calibration
+{
+template <typename T>
+void calculatePixelCoords(const T* const camera_intrinsics, const T* const camera_T_world,
+                          const T* const world_T_target, const T* const obj_point_in_target, T* const pixelCoords)
+{
+  T p[3];
+
+  ceres::AngleAxisRotatePoint(world_T_target, obj_point_in_target, p);
+  p[0] += world_T_target[3];
+  p[1] += world_T_target[4];
+  p[2] += world_T_target[5];
+
+  ceres::AngleAxisRotatePoint(camera_T_world, p, p);
+  p[0] += camera_T_world[3];
+  p[1] += camera_T_world[4];
+  p[2] += camera_T_world[5];
+
+  T x_prime = p[0] / p[2];
+  T x_prime_squared = x_prime * x_prime;
+  T y_prime = p[1] / p[2];
+  T y_prime_squared = y_prime * y_prime;
+
+  const T& fx = camera_intrinsics[0];
+  const T& fy = camera_intrinsics[1];
+  const T& cx = camera_intrinsics[2];
+  const T& cy = camera_intrinsics[3];
+  const T& k1 = camera_intrinsics[4];
+  const T& k2 = camera_intrinsics[5];
+  const T& p1 = camera_intrinsics[6];
+  const T& p2 = camera_intrinsics[7];
+  const T& k3 = camera_intrinsics[8];
+  // const T& k4 = camera_intrinsics[9];
```

Comment:
The `plumb_bob` distortion model uses only k1, k2, p1, p2 and k3.
There are apparently other models out there that can utilize the other parameters that I commented out but I have not actually found any of these other distortion models (feel free to let me know if you find any). 

Even the `sensor_msgs/CameraInfo.msg` message only mentions the `plum_bob` model (although it has the storage for other models if someone where to propose an alternate)

So the commenting out is a temporary fix until I code it more generally. I don't think it is a system-breaking situation either since all the cameras we test with use the `plumb_bob` model anyway.